Add hierarchical module names to torchFX graph.node #87659

alexmsettle opened the pull request:
Fixes #87659

Pass down the module hierarchy from module.named_modules() to the name field of graph.node.
This makes it so the name of each node contains descriptive information about the network architecture. 

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

alexmsettle committed:
Add hierarchical module names to torchFX graph.node #87659

alexmsettle committed:
Fixed lint errors

alexmsettle committed:
lint fix

alexmsettle committed:
Removed debug prints

alexmsettle committed:
Fixed failing regression test int test.ao.sparsity

alexmsettle committed:
Added optional param to function to match change

in the  base class function definition

ezyang commented:
@SherlockNoMad bounce this back to me if you're not the right reviewer

alexmsettle committed:
Fixed backward compatibility test issues

Moved implementation of Scope Context from
torch.ao.quantization.fx.QuantizationTracer to base class
torch.fx.Tracer so it is available to all the fx tracers.

SherlockNoMad commented on torch/fx/_symbolic_trace.py:
```diff
@@ -430,9 +489,29 @@ def call_module(
+            self.scope.module_path,
+            self.scope.module_type,
+        )
+        node.meta['nn_module_stack'] = {self.scope.module_path : self.scope.module_type}
```
This is different from what is store in nn_module_stack from the dynamo trace... 
Ideally, we want to make it consistent... 

node.meta['nn_module_stack'] is currently an OrderedDictionary that store the entire stack of modules. 

@jerryzh168, do you have a preference which structure it should take?

jerryzh168 replied:
I think today we actually only use a single level of scope (modue_fqn, module_type) in quantization (e.g. ("sub.conv", nn.Conv2d)) (actually the scope context code in this PR is the same as the ones defined in quantization tracer: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/tracer.py#L11), but I think theoretically we should be storing a stack of (module_fqn, module_type) 
e.g. 
for a module like following:
class Sub(...):
    def __init__(self, ...):
          self.conv = nn.Conv2d(...)

class M(...):
    def __init__(self, ...):
          self.sub = Sub()

     def forward(self, x):
           return self.sub(x)

for the conv aten op, the scope should be something like
[("sub.conv", nn.Conv2d), ("sub", Sub)] since to be precise, the conv aten node is in the scope of the Conv module and Sub module. We can get module fqn for sub just by looking at "sub.conv", but we can't get the module type of "Sub" easily if we just have a single entry of ("sub.conv", nn.Conv2d)

alexmsettle replied:
FYI - I actually just moved the Scope related code from quantization tracer into the base class Tracer so everyone has access to the module scope. 

Doesn't the module_fqn already account for the hierarchy?  Eg when the aten::conv is actually called the module_fqn would be sub.conv and the op would be aten::conv2d.  Ok now I see the issue, if you want to preserve the module type at each level of the hierarchy then you do need a listing of each level of hierarchy.

alexmsettle replied:
@SherlockNoMad  I just added an ordered dict for module stack.  Now the code writes an ordered dict to node.meta['nn_module_stack']

For the example code above w/ Module and Sub module I get this output  `Node sub_mod_conv_0 - mod {'sub_mod': "<class '__main__.SubMod'>", 'sub_mod.conv_0': "<class 'torch.nn.modules.conv.Conv2d'>"}`

alexmsettle committed:
Fixed lint errors

Fixed missing import of Scope class

alexmsettle committed:
Lint fix

alexmsettle committed:
Fixed backward compatibility test failures

alexmsettle committed:
Fixed CI regression error GraphAppendingTacer

alexmsettle committed:
Fixed error from GraphAppendTracer

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -13,6 +13,64 @@
+        self, scope: Scope, current_module: torch.nn.Module, current_module_path: str
+    ):
+        super().__init__()
+        self.prev_module_type = scope.module_type
+        self.prev_module_path = scope.module_path
+        self.scope = scope
+        self.scope.module_path = current_module_path
+        self.scope.module_type = type(current_module)
```
nit: I feel these can probably be simplified by just storing previous scope object and current scope object

alexmsettle replied:
I just fixed that.  ^^^

Working on capturing the scope stack now.

alexmsettle committed:
Fixed issue from code review

Added new class definitions to __all__ list

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle committed:
Added scope_stack for meta[nn_module_stack]

alexmsettle committed:
Saved the module stack in an ordered dict

to match dynamo #961 usage of node.meta['nn_module_stack']

alexmsettle commented:
ping @jerryzh168 @SherlockNoMad - I think I addressed all the issues raised in the code review.   Should I add a unit test to protect the module names?  

I had to move much of my changes to the TracerBase class because of backward compatibility testing issues.

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -11,7 +12,68 @@
+    """
+
+    def __init__(
+        self, scope: Scope, current_module: torch.nn.Module, current_module_path: str
```
nit: just to make this clearer, should we use
```
def __init__(self, prev_scope: Scope, current_scope: Scope):
```?

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -11,7 +12,68 @@
+        self, scope: Scope, current_module: torch.nn.Module, current_module_path: str
+    ):
+        super().__init__()
+        self._prev_scope = Scope("", None)
+        self._prev_scope.module_type = scope.module_type
+        self._prev_scope.module_path = scope.module_path
+        self.scope = scope
+        self.scope.module_path = current_module_path
+        self.scope.module_type = type(current_module)
```
this part feels confusing, we can probably cleanup this after we change the signature as well

alexmsettle replied:
Fixed this by changing the signature from previous comment.

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -11,7 +12,68 @@
+        return self.scope
+
+    def __exit__(self, *args):
+        self.scope.module_path = self._prev_scope.module_path
+        self.scope.module_type = self._prev_scope.module_type
```
feels like this could just be `self.scope = self._prev_scope`?

alexmsettle replied:
Fixed this too.

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle committed:
Made code review changes.

alexmsettle committed:
Added test case for nn_module_stack

alexmsettle commented:
Thanks for reviewing my changes @jerryzh168 .  I was on vacation last week, so only just got to it today.

I also just added a test case to protect the node.meta['nn_module_stack']

alexmsettle committed:
Fix CI errors

alexmsettle committed:
Cleaned up scope context manager

alexmsettle committed:
Removed references to Scope and ScopeContextManager

from ao test because these classes were moved
to the fx Tracer base

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle commented:
Hey @jerryzh168 @SherlockNoMad  I think these changes are ready to go.  Could you please review?

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -11,7 +12,71 @@
+    the current module, and when we exit, we'll restore the previous scope information.
+    """
+
+    def __init__(
+        self, scope: Scope,
+        current_module: torch.nn.Module,
+        current_module_path: str
+    ):
```
nit: I think it might be clearer if we just pass around the Scope object?

something like the following:

```
def __init__(
        self, 
        prev_scope: Scope,
        current_scope: Scope,
    ):
```

alexmsettle replied:
Done

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -43,7 +117,14 @@ def create_node(self, kind : str, target : Target,
 
-        return self.graph.create_node(kind, target, args, kwargs, name, type_expr)
+        node = self.graph.create_node(kind, target, args, kwargs, name, type_expr)
+        self.node_name_to_scope[node.name] = (
+            self.scope.module_path,
+            self.scope.module_type,
+        )
```
maybe we can add a TODO to remove this, it now becomes redundant since we have `node.meta["nn_module_stack"]`

jerryzh168 commented on torch/fx/_symbolic_trace.py:
```diff
@@ -250,6 +249,13 @@ def __init__(
+        # Maps the containing module's name to the operator name
+        self.scope = Scope("", None)
+        # Records the module call stack
+        self.module_stack = {}
```
should this be an OrderedDict?

jerryzh168 replied:
also nit: maybe rename this to `nn_module_stack` to make it clearer

alexmsettle replied:
Added typing.OrderedDict to initialization

Done

jerryzh168 commented on torch/fx/_symbolic_trace.py:
```diff
@@ -430,9 +436,17 @@ def call_module(
+        with ScopeContextManager(self.scope, m, module_qualified_name) as _scope:
+            # module_stack is an ordered dict so writing then deleting the
+            # entry is equivalent to push/pop on a list
+            self.module_stack[_scope.module_path] = str(_scope.module_type)
```
actually, it feels like we can probably maintain this thing in a scope as well

alexmsettle replied:
The self.module_stack is the data structure that populates node.meta['nn_module_stack'].  It is supposed to be an OrderedDict to be consistent with torch dynamo logic that @SherlockNoMad implemented.

alexmsettle committed:
Code review fixes

alexmsettle committed:
Merge remote-tracking branch 'origin/alex_torchfx' into alex_torchfx

Pulled in latest changes from pytorch master

alexmsettle commented on test/test_fx.py:
```diff
@@ -1679,6 +1679,36 @@ def forward(self, x):
+            mod_stack = node.meta.get('nn_module_stack', {})
+            if mod_stack:
+                break
+        stack_list = [{item: mod_stack[item]} for item in mod_stack]
```
Done

alexmsettle committed:
Fixed lint error OrderedDict

alexmsettle committed:
Using typing.Dict for ordered Dict

alexmsettle committed:
Used collections.OderedDict type for module_stack

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle commented:
Hey @jerryzh168  - I made the changes you requested and just synced with the main pytorch branch.  All tests are passing. Can you please review?

jerryzh168 commented on torch/fx/proxy.py:
```diff
@@ -1,17 +1,83 @@
+        super().__init__()
+        # Make a copy of scope to restore on context block exit
+        self._prev_scope = copy.copy(prev_scope)
+        # Modify the prev_scope that was passed in by reference
+        prev_scope.module_path = current_scope.module_path
+        prev_scope.module_type = current_scope.module_type
```
why do we have to do this?

jerryzh168 replied:
OK I looked again, it looks like `prev_scope` is actually a `scope` variable that keeps track of current scope, maybe we should update the name to reflect this better, e.g. update `prev_scope` to just `scope`

jerryzh168 replied:
I feel something like the following might be clearer (pretty similar to original code)
```
def __init__(
        self,
        scope: Scope,
        current_scope: Scope,
    ):
    # keep a copy of previous scope so that we can restore the scope
    self._prev_scope = copy.copy(scope)
    # update scope to current scope
    scope.module_path = current_scope.module_path
    scope.module_type = current_scope.module_type
    # holding a reference so that we can restore it
    self.scope = scope

...

def __exit__(self, *args):
        # restore the scope to previous scope
        self.scope.module_path = self._prev_scope.module_path
        self.scope.module_type = self._prev_scope.module_type
        return
```

alexmsettle replied:
You beat me to it.  Your comments above make sense.  Using the name prev_scope made the code confusing.

alexmsettle committed:
Code review fixes to ScopeContextManager

alexmsettle committed:
Pulled in changes from main pytorch branch

Merge remote-tracking branch 'origin/alex_torchfx' into alex_torchfx

alexmsettle commented:
Ok @jerryzh168 - I made the above changes and all tests are passing.  Is this change set ready for sign off?

alexmsettle committed:
Merge branch 'pytorch:master' into alex_torchfx

alexmsettle commented:
Ping @jerryzh168  - I just synced from the main upstream branch and the tests pass.  Could you pls take another look so we can get this merged?

jerryzh168 reviewed (approved):
LGTM, thanks for addressing all the comments!

jerryzh168 commented on torch/fx/_symbolic_trace.py:
```diff
@@ -250,6 +250,13 @@ def __init__(
+        # Records the module call stack
+        self.module_stack = collections.OrderedDict()
+        # Mapping of node name to module scope
+        self.node_name_to_scope: Dict[str, Tuple[str, type]] = {}
```
another thing we could do is just to attach the scope to node itself, but this will involve quantization code as well and we can do it in a separate PR

alexmsettle replied:
ok that sounds good.

alexmsettle commented:
@pytorchbot merge

alexmsettle commented:
> LGTM, thanks for addressing all the comments!

Thanks for all the feedback and reviews @jerryzh168
